### PR TITLE
Mechanism to orca backend in synch with Actual Network state Backend #72

### LINKED ISSUE #72: Mechanism to orca backend in synch with Actual Network state Backend
Use Case - 

1. User makes some changes on device (via cli for example) when the ORCA was down but device was already discovered in ORCA, when the ORCA comes up as device was already discovered in ORCA, a discovery will not be triggered hence the changes will not be visible in ORCA untill user rediscovers the device. 
2. User makes some changes on device (via cli for example) when the ORCA was up, but user make config changes for which gNMI subscription notifications are not supported. In this case also ORCA will not be able to get changes done in the config untill user rediscovers the device.

ORCA_BACKEND -

- [x] orca_backend should have a REST end point to save the rediscovery config in sqllite DB, read the config if found any schedule the discovery to be triggered according the interval set. 
- [x] Also separate feature discovery end point should be provided. i.e. discover_interface, discover_vlan etc. 
- [x] Any set operation should be blocked during any device discovery or feature discovery. If not done this way, Neo4J might remain in inconsistent state. A proper HTTP error message should be returned regarding why config is not possible right now.

### PR DESCRIPTION
**Pull request by rahul-1475**
Fixes #72  

- Added new apis to discover by feature.
- Added new apis to sxhedule discovery.
- Machinism to block put request when discovery or other config in progresss.
- Store system state and get state.
- Test cases for scheduling and state.

Note: Feature discovery has no test cases . Integration test done with ui

### CODE REVIEW

**Comment by rahul-1475** [comment]
orca_nw_lib pr: [#97](https://github.com/STORDIS/orca_nw_lib/pull/97)

File: state_manager/middleware.py [hunk 1]
```diff
@@ -0,0 +1,104 @@
+import datetime
```
Comment (kamalkrbh) [inline state_manager/middleware.py:1]: Unused import.

File: state_manager/test/test_state.py [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+    def tearDownClass(cls):
+        if scheduler.running:
+            scheduler.shutdown()
+
+    def test_discovery_state(self):
+        response = self.client.get(reverse("orca_state", kwargs={"device_ip": "127.0.0.1"}), )
+        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
+
+        request = self.factory.put(path=reverse('discover'), data={"address": "127.0.0.1"}, format="json")
+        middleware = BlockPutMiddleware(lambda req: discovery_stub(req, self.client))
```
Comment (kamalkrbh) [inline state_manager/test/test_state.py:66]: req object is unused in discover_stub function

File: state_manager/test/test_state.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.json().get("device_ip"), "127.0.0.1")
+        self.assertEqual(response.json().get("state"), str(State.AVAILABLE))
+
+    def test_config_state(self):
+        response = self.client.get(reverse("orca_state", kwargs={"device_ip": "127.0.0.1"}), )
+        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
+
+        request = self.factory.put(path=reverse('device_interface_list'), data={"mgt_ip": "127.0.0.1"}, format="json")
+        middleware = BlockPutMiddleware(lambda req: put_stub(req, self.client))
```
Comment (kamalkrbh) [inline state_manager/test/test_state.py:94]: req object is unused in put_stub object.

File: state_manager/test/test_state.py [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+        self.assertEqual(response.status_code, status.HTTP_200_OK)
+        self.assertEqual(response.json().get("device_ip"), "127.0.0.1")
+        self.assertEqual(response.json().get("state"), str(State.AVAILABLE))
+
+    def test_feature_discovery_state(self):
+        response = self.client.get(reverse("orca_state", kwargs={"device_ip": "127.0.0.1"}), )
+        self.assertEqual(response.status_code, status.HTTP_204_NO_CONTENT)
+
+        request = self.factory.put(path=reverse('discover_by_feature'), data={"mgt_ip": "127.0.0.1"}, format="json")
+        middleware = BlockPutMiddleware(lambda req: feature_discovery_stub(req, self.client))
```
Comment (kamalkrbh) [inline state_manager/test/test_state.py:80]: req object is unused in feature_discovery_stub

File: state_manager/middleware.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+            ip_next_state = self._get_device_state(request)
+            for ip, next_state in ip_next_state.items():
+                state_obj, created = OrcaState.objects.get_or_create(
+                    device_ip=ip,
+                    defaults={"state": str(State.AVAILABLE)},
+                )
+
+                # Mapping states to error messages
+                state_block_map = {
+                    str(State.DISCOVERY_IN_PROGRESS): "Discovery in progress",
```
Comment (kamalkrbh) [inline state_manager/middleware.py:30]: Enum string could be mapped in the enum definition.

**Pull request merged**